libhoth_spi.c

```diff
@@ -0,0 +1,286 @@
+        goto err_out;
+    }
+
+    dev = calloc(1, sizeof(struct libhoth_device));
+    if (dev == NULL) {
+        status = LIBHOTH_ERR_MALLOC_FAILED;
+        goto err_out;
+    }
+
+    spi_dev = calloc(1, sizeof(struct libhoth_spi_device));
+    if (spi_dev == NULL) {
+        status = LIBHOTH_ERR_MALLOC_FAILED;
+        goto err_out;
+    }
+
+    if(options->bits)
+    {
+        const uint8_t bits = (uint8_t)options->bits;
+        if (ioctl(fd, SPI_IOC_WR_BITS_PER_WORD, bits) < 0)
+        {
+            status = LIBHOTH_ERR_FAIL;
+            goto err_out;
+        }
+    }
+
+    if(options->mode)
+    {
+        const uint8_t mode = (uint8_t)options->mode;
+        if (ioctl(fd, SPI_IOC_WR_MODE, &mode) < 0)
+        {
+            status = LIBHOTH_ERR_FAIL;
+            goto err_out;
+        }
+    }
+
+    if(options->speed)
+    {
+        const uint32_t speed = (uint32_t)options->speed;
+        if (ioctl(fd, SPI_IOC_WR_MAX_SPEED_HZ, &speed) < 0)
+        {
+            status = LIBHOTH_ERR_FAIL;
+            goto err_out;
+        }
+    }
+
+    spi_dev->fd = fd;
+    spi_dev->mailbox_address = options->mailbox;
+    spi_dev->address_mode_4b = 1;
+
+    dev->send = libhoth_spi_send_request;
+    dev->receive = libhoth_spi_receive_response;
+    dev->close = libhoth_spi_close;
+    dev->user_ctx = spi_dev;
+
+    *out = dev;
+    return LIBHOTH_OK;
+
+    err_out:
+    if (dev != NULL) {
+        free(dev);
+    }
+    if(spi_dev != NULL){
+        close(fd);
+        free(spi_dev);
+    }
+
+    return status;
+}
+
+int libhoth_spi_send_request(struct libhoth_device* dev,
+                                const void* request, size_t request_size) {
+    if(dev == NULL) {
+        return LIBHOTH_ERR_INVALID_PARAMETER;
+    }
+
+    struct libhoth_spi_device* spi_dev = (struct libhoth_spi_device*)dev->user_ctx;
+    
+    return spi_nor_write(spi_dev->fd, spi_dev->address_mode_4b, spi_dev->mailbox_address, request, request_size);
+}
+
+int libhoth_spi_receive_response(struct libhoth_device* dev, void *response,
+                                size_t max_response_size, size_t* actual_size,
+                                int timeout_ms) {
+    if(dev == NULL) {
+        return LIBHOTH_ERR_INVALID_PARAMETER;
+    }
+
+    if(max_response_size < 8){
+        return LIBHOTH_ERR_INVALID_PARAMETER;
+    }
+
+    size_t total_bytes;
+    int status;
+    struct ec_host_response host_response;
+    struct libhoth_spi_device* spi_dev = (struct libhoth_spi_device*)dev->user_ctx;
+    
+    // Read Header From Mailbox
+    status = spi_nor_read(spi_dev->fd, spi_dev->address_mode_4b, spi_dev->mailbox_address, response, 8);
+    if(status != LIBHOTH_OK){
+        return status;
+    }
+
+    total_bytes = 8;
+    memcpy(&host_response, response, 8);
+    if(actual_size)
+        *actual_size = total_bytes;
+
+    if(max_response_size < (total_bytes + host_response.data_len)){
+        return LIBHOTH_ERR_INVALID_PARAMETER;
```

Comment:
This isn't the most descriptive error in the world.

Reply:
Added a new error code to better capture this error.